src/System.Private.CoreLib/shared/System/Number.BigInteger.cs

```diff
@@ -0,0 +1,749 @@
+
+            public void ShiftLeft(uint shift)
+            {
+                // Process blocks high to low so that we can safely process in place
+                var length = _length;
+
+                if ((length == 0) || (shift == 0))
+                {
+                    return;
+                }
+
+                uint blocksToShift = Math.DivRem(shift, 32, out uint remainingBitsToShift);
+
+                // Copy blocks from high to low
+                int readIndex = (length - 1);
+                int writeIndex = readIndex + (int)(blocksToShift);
+
+                // Check if the shift is block aligned
+                if (remainingBitsToShift == 0)
+                {
+                    Debug.Assert(writeIndex < MaxBlockCount);
+
+                    while (readIndex >= 0)
+                    {
+                        _blocks[writeIndex] = _blocks[readIndex];
+                        readIndex--;
+                        writeIndex--;
+                    }
+
+                    _length += (int)(blocksToShift);
+
+                    // Zero the remaining low blocks
+                    Buffer.ZeroMemory((byte*)(_blocks.GetPointer()), (blocksToShift * sizeof(uint)));
+                }
+                else
+                {
+                    // We need an extra block for the partial shift
+                    writeIndex++;
+                    Debug.Assert(writeIndex < MaxBlockCount);
+
+                    // Set the length to hold the shifted blocks
+                    _length = writeIndex + 1;
+
+                    // Output the initial blocks
+                    uint lowBitsShift = (32 - remainingBitsToShift);
+                    uint highBits = 0;
+                    uint block = _blocks[readIndex];
+                    uint lowBits = block >> (int)(lowBitsShift);
+                    while (readIndex > 0)
+                    {
+                        _blocks[writeIndex] = highBits | lowBits;
+                        highBits = block << (int)(remainingBitsToShift);
+
+                        --readIndex;
+                        --writeIndex;
+
+                        block = _blocks[readIndex];
+                        lowBits = block >> (int)lowBitsShift;
+                    }
+
+                    // Output the final blocks
+                    _blocks[writeIndex] = highBits | lowBits;
+                    _blocks[writeIndex - 1] = block << (int)(remainingBitsToShift);
+
+                    // Zero the remaining low blocks
+                    Buffer.ZeroMemory((byte*)(_blocks.GetPointer()), (blocksToShift * sizeof(uint)));
+
+                    // Check if the terminating block has no set bits
+                    if (_blocks[_length - 1] == 0)
+                    {
+                        _length--;
+                    }
+                }
+            }
+
+            [StructLayout(LayoutKind.Sequential, Size = (MaxBlockCount * sizeof(uint)))]
+            private struct BlocksBuffer
```

Comment:
`fixed uint _buffer[MaxBlockCount]` ?

Reply:
Won't that cause a perf hit due to `UnsafeValueTypeAttribute`?

Reply:
You can measure it. 

You are using unsafe stackallocated buffers, so you do want to pay for the GS cookie checks. I assume that the C/C++ code payed for them too.

Reply:
We should move to a C# 7.3 compiler version (we are still on 2.8.0-beta4 right now) so that we can index a fixed-buffer without pinning: https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-7-3#indexing-fixed-fields-does-not-require-pinning

Reply:
https://github.com/dotnet/buildtools/pull/2163

Reply:
This is blocked more generally on: https://github.com/dotnet/coreclr/pull/19878#issuecomment-422179390.